macros/src/main/scala/com/thaj/safe/string/interpolator/SafeString.scala

```diff
@@ -38,6 +43,24 @@ object SafeString {
         }
       }
 
+      def quasiquote(acc: c.universe.Tree, field: c.universe.Tree) = {
+        acc match {
+          case q"""StringContext.apply(..$raw).s(..$previousElements)""" => q"""StringContext.apply(..$raw).s(($previousElements :+ ..${field}) :_*)"""
+          case _ => q"""${acc}.s(..$field)"""
+        }
+      }
+
+      def typeOfElement(tree: Tree): ElementType = {
+        val tagType: Type = c.WeakTypeTag(tree.tpe).tpe
+
+        if(tagType =:= typeOf[String])
+          StringType
+        else if (!(tagType =:= typeOf[String]) && tagType.typeSymbol.isClass && tagType.typeSymbol.asClass.isCaseClass)
+          CaseClassType
+        else
+          UnknownType
+      }
```

Comment:
There is a double loop here.
1. Find ElementType
2. Use ElementType.

This could be one step while allowing readability using type that you envisioned.

Could you verify if we can pattern match on `tagType ` directly ?

Reply:
I am pretty sure using pattern match, we can find if tree is case class.
It will be helpful if you could add that :)

Reply:
Something like

```scala
classDecl match {
  case q"case class $className(..$fields) extends ..$parents { ..$body }" =>
    (className, fields, parents, body)
}

// where classDecl is of the type `ClassDef`. 
```